Add remote guardrails model support (#7377)
opensearch-project/documentation-website: _ml-commons-plugin/remote-models/guardrails.md

Guardrails are also triggered when a prompt matches the supplied regular expression.

## Validating input/output using a guardrail model
**Introduced 2.15**
{: .label .label-purple }

For more advanced validation, you can configure a guardrail model---a separate LLM whose purpose is to validate the user input and the LLM output. In this example, you'll configure two models:

> - Chat model: An Anthropic Claude model hosted on Amazon Bedrock, whose purpose is to hold a conversation with a user

Reviewer: ```suggestion
- Chat model: An Anthropic Claude model hosted on Amazon Bedrock whose purpose is to hold a conversation with a user.
```

> - Guardrail model: An OpenAI model, whose purpose is to validate the user input and Anthropic Claude output.

Reviewer: ```suggestion
- Guardrail model: An OpenAI model whose purpose is to validate the user input and Anthropic Claude output.
```

### Step 1: Create a connector for the guardrail model

> First, create a connector to the OpenAI guardrail model. Note that the OpenAI prompt instructs the model to respond only with the words `accept` or `reject`, depending on whether the input/output is acceptable. Additionally, the request contains the `response_filter` parameter that specifies the field in which the guardrail model will provide the validation result: 

Reviewer: ```suggestion
First, create a connector to the OpenAI guardrail model. Note that the OpenAI prompt instructs the model to respond only with the words `accept` or `reject`, depending on whether the input/output is acceptable. Additionally, the request contains the `response_filter` parameter, which specifies the field in which the guardrail model will provide the validation result: 
```

        "max_tokens": 7,
        "temperature": 0,
        "model": "gpt-3.5-turbo-instruct",
        "prompt": "You are a helpful assistant and an expert judge of content quality. Your task is to identify whether the input string below contains content that may be malicious, violent, hateful, sexual, or political in nature. Your answer should consist of a single word, either reject or accept. If the input belongs to any of these categories, please write reject. Otherwise, write accept. \\n\\nHere is the input: ${parameters.question}. \\n\\nYour answer: ",
        "response_filter": "$.choices[0].text"
    },
    "credential": {
        "openAI_key": "<openAI_key>"
    },
    "actions": [
        {
            "action_type": "predict",
            "method": "POST",
            "url": "https://${parameters.endpoint}/v1/completions",
            "headers": {
                "Authorization": "Bearer ${credential.openAI_key}"
            },
            "request_body": "{ \"model\": \"${parameters.model}\", \"prompt\": \"${parameters.prompt}\", \"max_tokens\": ${parameters.max_tokens}, \"temperature\": ${parameters.temperature} }"
        }
    ]
}
```
{% include copy-curl.html %}

> The response contains the connector ID that you'll use in the next steps:

Reviewer: ```suggestion
The response contains the connector ID used in the next steps:
```

```json
{
  "connector_id": "j3JVDZABNFJeYR3IVPRz"
}
```

### Step 2: Register a model group for the guardrail model

To register a model group for the OpenAI guardrail model, send the following request:

```json
POST /_plugins/_ml/model_groups/_register
{
    "name": "guardrail model group",
    "description": "This is a guardrail model group."
}
```
{% include copy-curl.html %}

> The response contains the model group ID that you'll use to register a model to this model group:

Reviewer: ```suggestion
The response contains the model group ID used to register a model to this model group:
```

```json
POST /_plugins/_ml/models/_register?deploy=true
{
    "name": "openai guardrails model",
    "function_name": "remote",
    "model_group_id": "ppSmpo8Bi-GZ0tf1i7cD",
    "description": "guardrails test model",
    "connector_id": "j3JVDZABNFJeYR3IVPRz"
}
```
{% include copy-curl.html %}

OpenSearch returns the task ID of the register operation and the model ID of the registered model:

```json
{
  "task_id": "onJaDZABNFJeYR3I2fQ1",
  "status": "CREATED",
  "model_id": "o3JaDZABNFJeYR3I2fRV"
}
```

To check the status of the operation, provide the task ID to the [Tasks API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/tasks-apis/get-task/):

```bash
GET /_plugins/_ml/tasks/onJaDZABNFJeYR3I2fQ1
```
{% include copy-curl.html %}

When the operation is complete, the state changes to `COMPLETED`:

```json
{
  "model_id": "o3JaDZABNFJeYR3I2fRV",
  "task_type": "DEPLOY_MODEL",
  "function_name": "REMOTE",
  "state": "COMPLETED",
  "worker_node": [
    "n-72khvBTBi3bnIIR8FTTw"
  ],
  "create_time": 1689793851077,
  "last_update_time": 1689793851101,
  "is_async": true
}
```

### Step 4 (Optional): Test the guardrail model

> You can test the guardrail model user input validation by sending requests that do an do not contain offensive words. 

Reviewer: ```suggestion
You can test the guardrail model user input validation by sending requests that do and do not contain offensive words. 
```

First, send a request that does not contain offensive words:

```json
POST /_plugins/_ml/models/o3JaDZABNFJeYR3I2fRV/_predict
{
  "parameters": {

>     "question": "how many indices"

Reviewer 2: ```suggestion
    "question": "how many indices do i have in my cluster"
```

```json
POST /_plugins/_ml/connectors/_create
{
  "name": "BedRock claude Connector",
  "description": "BedRock claude Connector",
  "version": 1,
  "protocol": "aws_sigv4",
  "parameters": {
      "region": "us-east-1",
      "service_name": "bedrock",
      "anthropic_version": "bedrock-2023-05-31",
      "endpoint": "bedrock.us-east-1.amazonaws.com",
      "auth": "Sig_V4",
      "content_type": "application/json",
      "max_tokens_to_sample": 8000,
      "temperature": 0.0001,
      "response_filter": "$.completion"
  },
  "credential": {
      "access_key": "<access_key>",
      "secret_key": "<secret_key>"
  },
  "actions": [
    {
      "action_type": "predict",
      "method": "POST",
      "url": "https://bedrock-runtime.us-east-1.amazonaws.com/model/anthropic.claude-v2/invoke",
      "headers": { 
        "content-type": "application/json",
        "x-amz-content-sha256": "required"
      },
      "request_body": "{\"prompt\":\"${parameters.prompt}\", \"max_tokens_to_sample\":${parameters.max_tokens_to_sample}, \"temperature\":${parameters.temperature},  \"anthropic_version\":\"${parameters.anthropic_version}\" }"
    }
  ]
}
```
{% include copy-curl.html %}

> The response contains the connector ID that you'll use in the next steps:

Reviewer: ```suggestion
The response contains the connector ID used in the next steps:
```

```json
{
  "connector_id": "xnJjDZABNFJeYR3IPvTO"
}
```

### Step 6: Register and deploy the chat model with guardrails

> To register and deploy the Anthropic Claude chat model, send the following request. Note that the `guardrails` object contains a `response_validation_regex` parameter, which specifies to only treat the input/output as valid if the guardrail model responds with a variant of the word `accept`:

Reviewer: ```suggestion
To register and deploy the Anthropic Claude chat model, send the following request. Note that the `guardrails` object contains a `response_validation_regex` parameter that specifies to only treat the input/output as valid if the guardrail model responds with a variant of the word `accept`:
```

```json
POST /_plugins/_ml/models/_register?deploy=true
{
    "name": "Bedrock Claude V2 model with openai guardrails model",
    "function_name": "remote",
    "model_group_id": "ppSmpo8Bi-GZ0tf1i7cD",
    "description": "Bedrock Claude V2 model with openai guardrails model",
    "connector_id": "xnJjDZABNFJeYR3IPvTO",
    "guardrails": {
        "input_guardrail": {
            "model_id": "o3JaDZABNFJeYR3I2fRV",
            "response_validation_regex": "^\\s*\"[Aa]ccept\"\\s*$"
        },
        "output_guardrail": {
            "model_id": "o3JaDZABNFJeYR3I2fRV",
            "response_validation_regex": "^\\s*\"[Aa]ccept\"\\s*$"
        },
        "type": "model"
    }
}
```
{% include copy-curl.html %}

OpenSearch returns the task ID of the register operation and the model ID of the registered model:

```json
{
  "task_id": "1nJnDZABNFJeYR3IvfRL",
  "status": "CREATED",
  "model_id": "43JqDZABNFJeYR3IQPQH"
}
```

### Step 7 (Optional): Test the chat model with guardrails

> You can test the Anthropic Claude chat model with guardrails by sending predict requests that do an do not contain offensive words. 

Reviewer: ```suggestion
You can test the Anthropic Claude chat model with guardrails by sending predict requests that do and do not contain offensive words. 
```
